File: 백준/Gold/9205. 맥주 마시면서 걸어가기/맥주 마시면서 걸어가기.py
# 노드들 간의 거리가 1000m를 넘어서는 안됨! (박스에 맥주를 최대 20개까지 담을 수 있고 50m에 맥주 한 병씩을 마셔야 하기 때문)
from collections import deque
import sys
input = sys.stdin.readline

def BFS():
    q = deque()
    q.append([home[0], home[1]])
    # q가 빌 때까지 반복
    while q:
        x, y = q.popleft()
        # 절댓값 함수 abs
        if abs(fest[0] - x) + abs(fest[1] - y) <= 1000:
            print("happy")
            return
        for i in range(n):
            if not visited[i]:
                store_x, store_y = store[i]
                if abs(store_x - x) + abs(store_y - y) <= 1000:
                    q.append([store_x, store_y])
                    visited[i] = 1
    print("sad")
    return

t = int(input())
for _ in range(t):
    # 맥주를 파는 편의점의 개수
    n = int(input())
    # 상근이네 집 x, y 좌표
    home = [int(x) for x in input().split()]
    store = []
    # 편의점의 위치
    for _ in range(n):
        x, y = map(int, input().split())
        store.append([x, y])
    # 펜타포트 락 페스티벌 위치
    fest = [int(x) for x in input().split()]
    # 편의점 방문 여부 체크하기 위한 리스트
    visited = [0 for _ in range(n)]
    BFS()

File: 백준/Gold/9205. 맥주 마시면서 걸어가기/맥주 마시면서 걸어가기.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.List;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
	static int n;
	static int[] home, fest;
	static List<int[]> store;
	static boolean[] visited;

	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;
		int T = Integer.parseInt(br.readLine());
		for(int t = 0; t < T; t++) {
			n = Integer.parseInt(br.readLine());
			home = new int[2];
			fest = new int[2];
			store = new ArrayList<int[]>();
			
			st = new StringTokenizer(br.readLine());
			home[0] = Integer.parseInt(st.nextToken());
			home[1] = Integer.parseInt(st.nextToken());
			
			// 편의점 좌표값 넣기
			for(int i = 0; i < n; i++) {
				st = new StringTokenizer(br.readLine());
				int x = Integer.parseInt(st.nextToken());
				int y = Integer.parseInt(st.nextToken());
				store.add(new int[] {x, y});
			}
			
			st = new StringTokenizer(br.readLine());
			fest[0] = Integer.parseInt(st.nextToken());
			fest[1] = Integer.parseInt(st.nextToken());
			
			if(distance(home, fest) <= 1000){
				System.out.println("happy");
				continue;
			}
			visited = new boolean[n];
			bfs();
		}
	}
	
	public static void bfs() {
		Queue<int[]> q = new ArrayDeque<int[]>();
		q.offer(home);
		
		while(!q.isEmpty()) {
			int[] a = q.poll();
			if(distance(a, fest) <= 1000) {
				System.out.println("happy");
				return;
			}
			for(int i = 0; i < n; i++) {
				if(!visited[i]) {
					int x = store.get(i)[0];
					int y = store.get(i)[1];
					if(distance(a, new int[] {x, y}) <= 1000) {
						q.offer(new int[] {x, y});
						visited[i] = true;
					}
				}
			}
		}
		System.out.println("sad");
		return;
	}
	
	public static int distance(int[] a, int[] b) {
		return Math.abs(a[0]-b[0]) + Math.abs(a[1]-b[1]);
	}
}